メモアプリ

ham-cap opened the pull request:
メモアプリの課題を提出いたします。
よろしくお願いいたします。

ham-cap committed:
メモの新規作成機能を実装

新規作成用のフォームからタイトルと本文を受け取り、作成日時と併せて
json形式でmemosディレクトリに保存する仕様とした。
ひとつのメモにつき1ファイルが作成される。
ファイル名をUUIDとし、以降既存メモの検索等で使用する。

ham-cap committed:
TOP画面に既存メモのタイトルが一覧表示されるようにした

各タイトルはそれぞれのメモデータの編集画面へとべるようにするために
リンクとしている。

ham-cap committed:
既存メモの編集機能を実装

既存メモの内容をフォームの初期値としている。

ham-cap committed:
メモの削除機能を実装

削除機能を実装した。
view側からMethodOverrideでPOSTをDELETEに換えて送信している。

ham-cap committed:
memo_app.rb内の重複している処理をメソッドとして切り出し・一部のviewに「TOPへ戻る」ボタンを追加

メモの詳細画面を表示する場合と編集画面を表示する場合の処理はその大部分が同じだったため
メソッドとして切り出して記述量を減らした。
また、複数のviewにTOP画面へ戻るためのリンクが無かったため追加した。

ham-cap committed:
READMEを編集・layout.erbの不要部分を削除

READMEが空白のままだったため、ローカルで動作させる際の手順を記載した。

ham-cap committed:
rubocopによる指摘を反映

主な変更点は以下の2点
・メモデータを読み込む際のメソッドをJSON.loadではなくJSON.parseに変更
・取得したメモデータを作成日順に並べ替える際に使用するメソッドをsortからsort_byに変更

ham-cap committed:
不要なファイルを削除

publicに不要なファイルが残っていたため削除

yuuu commented on .byebug_history:
```diff
@@ -0,0 +1,22 @@
+quit
```
こちらはリポジトリには不要なファイルだと思われますので、`.gitignore` を使ってコミット対象から外すようにしましょう。

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,96 @@
+def collect_memo_data
+  file_names = Dir.glob('*', base: './memos')
+  @memo_files = []
+  file_names.each do |name|
```
ここは `file_names` から `@memo_files` を作っている箇所ですので `each` ではなく `map` を使った方がスッキリ書けますね。

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,96 @@
+def find_selected_memo(number)
+  index = number.to_i - 1
+  selected_memo = @memo_files[index]
+  @selected_number = selected_memo['number']
```
1属性1変数とすると、属性が増えた場合に変数も増えてしまいメンテナンスが大変ですね。
`@selected_memo` をそのまま `erb` で使うようにするのはいかがでしょうか？

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,96 @@
+  erb :created
+end
+
+get '/memos/show/:number' do |number|
```
リソース名を表す、という意味では `/memos/:number`というパスが良さそうですね

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,96 @@
+  erb :show
+end
+
+get '/memos/edit/:number' do |number|
```
こちらもリソースのeditformという意味では `/memos/:number/edit` の方が良いですね

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,96 @@
+    read_line = file.read
+    JSON.parse(read_line)
+  end
+  @original_file['title'] = CGI.escapeHTML(params[:memo_title])
```
保存するデータ自体をescapeすると、このメモを再度編集する際に入力した内容が変わっていて戸惑う懸念があります。保存するときではなく描画するとき(erb)にてescapeするようにしましょう。

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,96 @@
+  collect_memo_data
+  sort_by_created_at
+  reverse_memos_order
+  @memo_titles = []
```
このインスタンス変数はこのブロックでしか使っていないようですね。不要であれば削除をお願いします。

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,96 @@
+  number_of_files = Dir.glob('*', base: './memos').size
+  @latest_number = number_of_files + 1
+  File.open("./memos/memo_#{@latest_number}.json", 'w') do |file|
+    JSON.dump({ 'id' => id, 'number' => @latest_number, 'title' => @title, 'body' => @body, 'created_at' => @created_at }, file)
```
Hashのkeyをどうしても文字列としたいケースを除いて、省略した記法を使うのが一般的ですね。

```ruby
JSON.dump({ id: id, number: @latest_number, title: @title, body: @body, created_at: @created_at }, file)
```

ham-cap committed:
collect_memo_dataメソッド内で使用するメソッドをeachからmapへ変更

file_namesから@memo_filesを作成する際、eachによる繰り返し処理ではなくmapを使用する方が
すっきりしてわかりやすくなるため変更した

ham-cap committed:
find_selected_memoメソッド内でメモのタイトルや本文を個別の変数に格納していたのをやめてerb側で@selected_memoから必要な属性を呼び出す仕様にした

各属性を個別の変数に格納して運用する場合、属性が増えたときのメンテナンスが大変になるため。

ham-cap committed:
メモを作成する際のハッシュの記法を省略記法に変更

キーを文字列で記述して=>で値と繋ぐ記法を使っていたが、省略した記法のほうが一般的なため

ham-cap committed:
メモの詳細画面と編集画面に移動する際のURLを変更

リソース名を指すものと考えると適切ではなかったため、詳細画面への遷移からはshowの部分を削除
編集画面への遷移は/memos/:number/editへ変更した

ham-cap committed:
サニタイジングを行うタイミングをデータの保存時ではなくviewでの描画時に変更

CGI.escapeHTMLを用いてメモの保存時にサニタイジングを行うと、入力されたデータ自体が書き換えられるため
後から再度編集する際に、最初に入力されたデータと違ってしまう。
そのため、保存されるデータ自体は入力時のままで、ブラウザに描画される際にサニタイジングを行うことで
これを回避する。

ham-cap commented:
レビューありがとうございます！
ご指摘いただいた箇所を修正いたしましたのでご確認お願いいたします。

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,88 @@
+
+post '/memos' do
+  id = SecureRandom.uuid
+  @title = params[:memo_title]#CGI.escapeHTML(params[:memo_title])
```
コメント部分は不要だと思われますので削除しておきましょう

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,88 @@
+end
+
+def find_selected_memo(number)
+  index = number.to_i - 1
```
Indexで参照すると、Memoを削除したときにindexがずれてしまい、同じURLであっても違うMemoを読み出すことにならないでしょうか？

yuuu commented on views/new.erb:
```diff
@@ -0,0 +1,7 @@
+<form action="/memos" method="POST">
+  <p><input name="memo_title" type=text></p>
+  <p><textarea name="memo_body" type=text></textarea></p>
+    <input type="submit" value="保存">
```
インデントがずれているようなので、揃えるようにしてください。

yuuu commented on memo_app.rb:
```diff
@@ -0,0 +1,88 @@
+end
+
+patch '/memos/:file_number' do
+  @original_file = File.open("./memos/memo_#{params[:file_number]}.json") do |file|
```
URLの一部がそのままファイル名になっているので `/memos//../../../var/www` のように `file_number` の部分にパスを埋め込まれると意図しないファイルにアクセスされる懸念があります。

Sinatraでもある程度対策されているようなので修正必須ではありませんが、以下URLを参照してどのようなリスクがあるのか、対策ができそうか検討をお願いします。
https://www.ipa.go.jp/security/vuln/websecurity-HTML-1_3.html

ham-cap committed:
find_selected_memoメソッドでの選択されたメモの参照方法を変更・new.erbのインデントずれを修正

選択されたメモを特定する際にインデックスを使って特定・参照していたが
既存のメモが削除された場合にメモの総数が変わり、インデックスもずれてしまうことから
各メモに持たせたnumberの値を使って特定するように変更した。

ham-cap committed:
メモのファイル名を数字からidに変更し、URLの一部とファイル名が一致しないよう変更・一度使ったnumberの値を再利用できないように変更

URLの末尾につく番号をファイル名に一致させて管理していたが、意図しないファイルへのアクセスを
防ぐために、ファイル名をUUIDにした。
また、新規メモの作成や削除を繰り返してもnumberの値が重複しないように、一度使ったnumberを
used_number.txtに記録する仕様に変更した。

ham-cap commented:
遅くなりましたが、コードを修正いたしましたのでご確認願います！

yuuu commented on memo_app.rb:
```diff
@@ -43,12 +43,25 @@ def find_selected_memo(number)
+    @original_array = file.readlines.map(&:to_i)
+  end
+  File.open('used_number.txt', 'w') do |file|
+    if @original_array.size.zero?
```
`@original_array.empty?` の方が短く直感的に書けますね

yuuu commented on memo_app.rb:
```diff
@@ -43,12 +43,25 @@ def find_selected_memo(number)
+    if @original_array.size.zero?
+      file.puts(1)
+    else
+      @latest_number = @original_array.max + 1
```
過去に使った番号を全て記述していくと、ファイルが肥大化していくので、最後に使った番号だけを記述した方が良さそうですね。その方が処理もシンプルになると思いますので修正を検討してみてください。

yuuu commented on memo_app.rb:
```diff
@@ -43,12 +43,25 @@ def find_selected_memo(number)
+    end
+  end
+  File.open("./memos/#{id}.json", 'w+') do |file|
+    File.open('used_number.txt') do |used_number|
```
ここですが再度読み直す必要があるのでしょうか？上の処理で確実に `@latest_number` がセットされるようにすれば、この `File.open` は省略できるはずです。

ham-cap committed:
used_number.txtの中身の有無を調べる際のメソッドを変更・used_number.txtに最新の番号のみを記録する仕様に変更・used_number.txtの無駄な読み込みを削除

used_number.txtに番号が記録されているかどうかを調べる際のメソッドをシンプルなものに変更した。
used_number.txtが肥大化するのを防ぐため、最新の番号のみを記録することにした。
メモのデータを作成する際に再度used_number.txtを読み込んでいたが、直前のif文でどちらに分岐しても
@latest_numberが定義されるようにすることで再読み込みを不要とした。

ham-cap commented:
修正いたしました！
ご確認お願いいたします！

yuuu commented on memo_app.rb:
```diff
@@ -47,21 +47,18 @@ def find_selected_memo(number)
       file.puts(@latest_number)
     end
   end
   File.open("./memos/#{id}.json", 'w+') do |file|
```
読み出しがなくなったので `mode`  は `'w'` で良いですね。

ham-cap committed:
新規メモ作成の際、File.openのモードをw+からwに変更

used_number.txtの再読み込みをなくし、書き込みのみできればよくなったため変更。
